### ISSUE TITLE
Add support for save/restore dock area layout

### ISSUE THREAD

**Issue by sccolbert** [phosphorjs/phosphor#33]
This should be independent of the current `addWidget` api.

**Comment by mladen107** [comment]
Hello,
this feature is really critical for project that I'm working on, is there any estimation about how long it should take to implement this? I see that you are refactoring dock panel at the moment. Should this refactoring help on adding support for this feature?

Great job with this project so far!

**Comment by jjrv** [comment]
What's the official way to read and programmatically change the DockPanel handle positions? An API like this would be nice:

First assume we set up the panel like so:

```
+-------+-----+-+
|       |     | |
|       |     | |
|       +-----+-+
|       |       |
+-------+-------+
```

So it's split like this:
- Horizontal 50% - 50%
  - Panel
  - Vertical 66% - 33%
    - Horizontal 75% - 25%
      - Panel
      - Panel
    - Panel

That's easy enough to represent in JSON. One possibility:

`[1.0, 0.5, [0.5, [0.66, 0.75, 0.25], 0.33]]`

Each branch in the tree is `[mySize, children...]` and leaf is `mySize`. Here `mySize` is always a number and `children` can be branches or leaves.

The exact structure (whether initial split is horizontal and vertical, and how many children at each split) is encoded in how `addWidget` has been called.

If necessary, a slightly more elaborate format could also encode which widgets are at each node, to allow batch-adding or reconfiguring everything through a single JavaScript object. Here's an example (again describing the ASCII art above):

```JSON
{
  "size": 1.0,
  "direction": "horizontal",
  "split": [
    {
      "size": 0.5,
      "widget": myLeftWidget
    }, {
      "size": 0.5,
      // "direction": "vertical",
      "split": [
        {
          "size": 0.66,
          // "direction": "horizontal",
          "split": [
            {
              "size": 0.75,
              "widget": myTopLeftWidget
            },
            {
              "size": 0.25,
              "widget": myTopRightWidget
            }
          ]
        },
        {
          "size": 0.33,
          "widget": myBottomWidget
        },
      ]
    }
  ]
}
```

The direction of each nested `split` is implicit, it's always alternating between `horizontal` and `vertical`.

**Comment by jjrv** [comment]
If you're not interested in implementing anything like this, I'll be happy to write a separate `npm` package to handle this. What would be the best way to query and set handle positions after adding widgets?

**Comment by sccolbert** [comment]
I definitely have intent to implement this. Along with maximize/tear-out feature, it will be the feature I work on after releasing 1.0.

**Comment by jjrv** [comment]
@sccolbert Any pointers for how to set the handle positions in the current version? Earlier accessing `setHandlePosition` inside nested SplitPanels used to work but through some private members.

I can probably get a POC done fast, and will have to do partially my own implementation anyway to support also serializing floating dialogs and nested dock panels inside them.

**Comment by sccolbert** [comment]
There is not public way to do that right now.

**Comment by jjrv** [comment]
Hm. I rather need at least something now, so here goes. Docs are in my long message above, `SizeData` follows the `[1.0, 0.5, [0.5, [0.66, 0.75, 0.25], 0.33]]` format.

```TypeScript
interface SizeData extends Array<SizeData | number> {
	0: number;
	1: number | SizeData;
	2: number | SizeData;
}

const sizes: SizeData = [1.0, 0.25, [0.75, [0.67, 0.75, 0.25], 0.33]];

function setSizes(dockPanel: widgets.DockPanel, sizes: SizeData) {
	function rec(splitLayout: any, [me, ...children]: SizeData) {
		let childLayout: any;
		let num = 0;

		for(let child of splitLayout.children) {
			if(child instanceof splitLayout.constructor) {
				childLayout = child;
			}
		}

		for(let child of children) {
			if(typeof(child) != 'number') {
				rec(childLayout, child)
				child = child[0];
			}
			splitLayout.sizers[num].sizeHint = child;
			++num;
		}

		splitLayout.normalized = true;
	}

	rec((dockPanel.layout as any)._root, sizes);
	dockPanel.update();
}

setSizes(dockPanel, sizes);
```

**Comment by sccolbert** [comment]
@jjrv As I said, this feature is the next thing I'll be working on. I won't be accepting PRs for it at this time.

**Comment by GordonSmith** [comment]
@sccolbert Related Questions - 
* Is it possible to then "lock" the layouts:  disable the ability to drag widgets from one pane to another - but resizing is ok?
* Optionally hide the tabs (most useful when there is only 1 tab - but also useful if the tabs can still be selected via API)

Here is my motivation:  I would like to use phosphor as a design time tool to let "managers" define the layout (they would have full access to tabs / dragging and dropping etc).  Once the layout is baked (via this proposed serialization), then the regular users would only have access to that "fixed" layout.  In this case there would be no need for "single" tabs (optional).

**Comment by mladen107** [comment]
I have implemented "proof of concept" solution based private API and `addWidget ` api. I have used for 6 months and it works but still I marked it as temp solution until official one arrives.

It is done in es6. I have some additional code for async loading of module widgets and also I have assumption that all my leaf widgets are type of special `ModuleWidget`(defined by it's `moduleDefinition `) and I'm using react so I'm passing react context.

Here is code for serializer and deserializer :

```
import {DockPanel} from 'phosphor/lib/ui/dockpanel';
import {TabPanel} from 'phosphor/lib/ui/tabpanel';
import {SplitPanel} from 'phosphor/lib/ui/splitpanel';
import ModuleWidget from 'extensions/phosphor/ModuleWidget';

function serialize(widget) {
    const result = {};
    let children,i;
    switch (widget.constructor) {
        case DockPanel:
            result.type = 'DockPanel';
            children = widget.layout.widgets._array;
            result.widgets = [];
            for (i = 0; i < children.length; i++) {
                result.widgets.push(serialize(children[i]));
            }
            return result;
        case TabPanel:
            children = widget.widgets._array;
            result.widgets = [];
            for (i = 0; i < children.length; i++) {
                result.widgets.push(serialize(children[i]));
            }
            result.type = 'TabPanel';
            return result;
        case SplitPanel:
            result.orientation = widget.orientation;
            result.widgets= [];
            result.type = 'SplitPanel';
            children = widget.widgets._array;
            result.relativeSizes = widget.layout.relativeSizes();
            for (i = 0; i < children.length; i++) {
                result.widgets.push(serialize(children[i]));
            }
            return result;
        case ModuleWidget:
            result.type = 'ModuleWidget';
            result.moduleDefinition = widget.moduleDefinition;
            return result;
        default:
            console.error("Unknown type " + widget.constructor.name, widget);
            result.type = 'UNKNOWN';
            return result;
    }
}

export default serialize;
```

```
import {DockPanel} from 'phosphor/lib/ui/dockpanel';
import ModuleWidget from 'extensions/phosphor/ModuleWidget';
import {moduleWidgetFactory} from 'modules/moduleLoader';

function deserialize(dockPanelData, reactContext, done) {
    if (!dockPanelData) {
        throw Error("dockPanelData is empty")
    }
    if (dockPanelData.type !== 'DockPanel') {
        throw Error("Root type empty or invalid. It must be DockPanel");
    }

    const dockPanel = new DockPanel();

    function createWidget(moduleDefinition, placement, ref, done) {
        moduleWidgetFactory(moduleDefinition, reactContext, null, (widget)=> {
            dockPanel.addWidget(widget, {mode: placement, ref: ref});
            done(widget);
        }, null);
    }

    function createTabWidgets(node, placement, ref, done) {
        const children = node.widgets;
        if (!children) {
            done(ref);
            return;
        }
        createWidget(children[0].moduleDefinition, placement, ref, (scopeRef)=> {
            if (children.length === 1) {
                done(scopeRef);
                return;
            }
            let i = 1;

            function loopFunction(ref) {
                createWidget(children[i].moduleDefinition, 'tab-after', ref, (ref)=> {
                    i++;
                    if (i < children.length) {
                        loopFunction(ref);
                    } else {
                        done(scopeRef);
                    }
                });
            }

            loopFunction(scopeRef);
        });
    }

    function generateNode(node, placement, ref, pivot, done) {
        if (node.type === 'TabPanel') {
            if (node !== pivot) {
                createTabWidgets(node, placement, ref, ()=> {
                    done();
                });
                return;
            } else {
                done();
                return;
            }
        }

        if (pivot) {
            let siblingsPlacement = (node.orientation === 'horizontal') ? 'split-right' : 'split-bottom';
            generateNodeSiblings(subarray(node.widgets, 1, node.widgets.length), siblingsPlacement, ref, ()=> {
                generateNode(node.widgets[0], undefined, ref, pivot, done)
            });
        } else {
            pivot = findFirstTabChild(node);
            createTabWidgets(pivot, placement, ref, (newRef)=> {
                let siblingsPlacement = (node.orientation === 'horizontal') ? 'split-right' : 'split-bottom';
                generateNodeSiblings(subarray(node.widgets, 1, node.widgets.length), siblingsPlacement, newRef, ()=> {
                    generateNode(node.widgets[0], undefined, newRef, pivot, done)
                });
            })
        }
    }

    function generateNodeSiblings(nodes, placement, ref, done) {
        if (nodes.length === 0) {
            done();
            return;
        }
        let i = nodes.length - 1;

        function loopFunction() {
            generateNode(nodes[i], placement, ref, null, ()=> {
                i--;
                if (i >= 0) {
                    loopFunction();
                } else {
                    done();
                }
            });
        }

        loopFunction();
    }

    if (dockPanelData.widgets.length > 0) {
        generateNode(dockPanelData.widgets[0], undefined, undefined, null, ()=> {
            adjustSizing(dockPanelData, dockPanel);
            done(dockPanel);
        });
    } else {
        done(dockPanel)
    }
}

function findFirstTabChild(node) {
    if (node.type === 'TabPanel') {
        return node;
    } else {
        return findFirstTabChild(node.widgets[0]);
    }
}


function subarray(array, start, end) {
    if (end >= start) {
        return array.slice(start, end);
    } else {
        return []
    }
}


function adjustSizing(data, widget) {
    let i;
    let childrenData;
    let childrenWidgets;
    switch (data.type) {
        case 'DockPanel':
            childrenData = data.widgets;
            childrenWidgets = widget.layout.widgets._array;
            for (i = 0; i < childrenData.length; i++) {
                adjustSizing(childrenData[i], childrenWidgets[i])
            }
            return;
        case 'SplitPanel':
            childrenData = data.widgets;
            childrenWidgets = widget.widgets._array;
            widget.layout.setRelativeSizes(data.relativeSizes);
            for (i = 0; i < childrenData.length; i++) {
                adjustSizing(childrenData[i], childrenWidgets[i])
            }
            return;
        case 'TabPanel':
            childrenData = data.widgets;
            childrenWidgets = widget.widgets._array;
            for (i = 0; i < childrenData.length; i++) {
                adjustSizing(childrenData[i], childrenWidgets[i])
            }
            return;
        case 'ModuleWidget':
            return;
        default:
            console.error("Unknown type for adjust sizing:" + node.type);
            return;
    }

}

export default deserialize;
```

**Comment by sccolbert** [comment]
@GordonSmith Sounds like you're describing a dashboarding tool. That's on our list of stuff to do this year in a formal way.

**Comment by sccolbert** [comment]
@mladen107 Your solution will no longer work since the private API has changed in the last few months.

**Comment by GordonSmith** [comment]
@sccolbert You are 100% correct - I already have an OSS framework for hosting and dashboarding charts (with a consistent + serializable) API wrapper, it includes a drag and drop auto layout engine similar to gridster (but the framework has no core dependencies) and I wanted to add tear away tabs - hence I landed here.  
FWIW the demo page is here: http://rawgit.com/hpcc-systems/Visualization/master/demos/dermatology.html?src/layout/Grid.complex&designMode=false
With some simple examples here:  https://bl.ocks.org/GordonSmith
All the runtime properties and layouts are serializable and the charts themselves are grouped into 5 categories which makes them inter-changeable.
Further - I have a new branch where I am in the process of converting it to TypeScript, with the hard part completed.  
I only mention the above as it would be pretty trivial to make a working dashboard demo with phosphor (when the time comes).

**Comment by sccolbert** [comment]
@GordonSmith I really like the dashboard editor you have, and it's very similar to some ideas I have floating around in my head. Phosphor already has a GridPanel layout which just needs to be ported to the new base classes (https://github.com/phosphorjs/phosphor-gridpanel), it wouldn't be a big stretch to add some interactivity to it. 

One thing that's been making me hesitant to use DockPanel for dashboards is that it might be *too* free form for this use case, since the widgets do not necessarily align to row/col boundaries. We could try to fix that by adding an additional mode to the dock panel which makes phantom row/col boundaries, but I haven't yet convinced myself that's a good idea.

What are thoughts for the interaction of tear-away tabs? I could implement that as either "drag outside the browser window", or make it a button on the tab bar.

**Comment by GordonSmith** [comment]
By "tear away tabs" I meant the ability to move tabs from one pane to a different pane and (in my world) create a new tabbed pane by dragging a widget on top of an existing widget.  
I didn't mean having the ability of popping out into a new browser window (like https://www.golden-layout.com/ does) - which IMO just opens a usability can of worms.

**Comment by sccolbert** [comment]
Ah, so your case would be covered by having non-tabbed areas in the dockpanel, and snapping to grid boundaries?

**Comment by GordonSmith** [comment]
Yes - In my mind (and I am not familiar with phosphor API - just the demo pages):
* In design mode:  Every pane has a tab and it works like the demo now - having a "snap to grid" feature with a configurable grid size is an added bonus)
* In run-time mode:  
    - Would disable all the layout editing - except the resizing of the splitter bars (optionally?) 
    - Optionally hide "single" tabs (or provide an API to do so - could be part of the serialization string?).

You should be able to end up with something like this easily then:  http://rawgit.com/hpcc-systems/Visualization/v1.10.0/demos/DataAnalyst/index.html 
(click the generate dashboard button, to create an interactive dashboard!)

FYI In my framework, to "publish" a property so that it is available in the editor and auto serializable you add the following code:
https://github.com/hpcc-systems/Visualization/blob/master/src/layout/Grid.js#L22-L35

In the TS version I am dabbling with annotations to achieve something similar:  https://github.com/GordonSmith/Visualization/blob/TSCWIP/src/common/Text.ts#L15-L26
(But I am still new to TS)

**Comment by sccolbert** [comment]
For anyone who's needs are not addressed by #222, would you please open a new issue?
